packages/provider/src/json-rpc-provider.ts

```diff
@@ -66,9 +80,15 @@ export class JsonRpcProvider implements AbstractProvider {
         'Content-Type': 'application/json',
       },
       body: JSON.stringify(body),
+    }).catch((error) => {
+      if (shouldRetryOnFailure) {
+        return performRetry()
+      } else {
+        throw error
+      }
     })
 
-    return rpcResponse
+    return (!rpcResponse.ok && shouldRetryOnFailure) ? performRetry() : rpcResponse
```

Comment:
Let's leave a comment here explaining what the reasoning is; I don't think it's immediately obvious that a `fetch` call can not throw but still pass non-OK responses (`400`, `500`, etc)